### PR TITLE
WIP prototype autograd and multiple dispatch!

### LINKED ISSUE #4: Fix several corner cases (in branch exp-space)
Branch: exp-space

## When output is empty
```julia
julia> einsum(((1, 2), (2, 1)), (a, a), ())
ERROR: MethodError: Base.reduce_empty(::typeof(Base.add_sum), ::Type{Union{}}) is ambiguous. Candidates:
  reduce_empty(::typeof(Base.add_sum), ::Type{T}) where T<:Union{UInt16, UInt32, UInt8} in Base at reduce.jl:236
  reduce_empty(::typeof(Base.add_sum), ::Type{T}) where T<:Union{Int16, Int32, Int8} in Base at reduce.jl:235
Possible fix, define
  reduce_empty(::typeof(Base.add_sum), ::Type{Union{}})
Stacktrace:
 [1] mapreduce_empty(::typeof(identity), ::Function, ::Type) at ./reduce.jl:251
 [2] mapreduce_empty_iter(::Function, ::Function, ::Tuple{}, ::Base.HasEltype) at ./reduce.jl:258
 [3] mapfoldl_impl at ./reduce.jl:57 [inlined]
 [4] #mapfoldl#187 at ./reduce.jl:72 [inlined]
 [5] mapfoldl at ./reduce.jl:72 [inlined]
 [6] #mapreduce#191 at ./reduce.jl:205 [inlined]
 [7] mapreduce(::Function, ::Function, ::Tuple{}) at ./reduce.jl:205
 [8] sum(::Function, ::Tuple{}) at ./reduce.jl:399
 [9] sum(::Tuple{}) at ./reduce.jl:416
 [10] macro expansion at ./multidimensional.jl:1337 [inlined]
 [11] permutedims!(::Array{Float64,0}, ::Array{Float64,0}, ::Tuple{}) at ./multidimensional.jl:1328
 [12] permutedims(::Array{Float64,0}, ::Tuple{}) at ./multidimensional.jl:1312
 [13] einsum(::Tuple{Tuple{Int64,Int64},Tuple{Int64,Int64}}, ::Tuple{Array{Float64,2},Array{Float64,2}}, ::Tuple{}) at /home/leo/jcode/OMEinsum.jl/src/einsum.jl:49
 [14] top-level scope at none:0
```

## When summing over one of the indices
```julia
julia> einsum(((1, 2), (2, 3)), (a, a), (1,))
ERROR: ArgumentError: no valid permutation of dimensions
Stacktrace:
 [1] permutedims(::Array{Float64,2}, ::Tuple{Int64}) at ./multidimensional.jl:1309
 [2] einsum(::Tuple{Tuple{Int64,Int64},Tuple{Int64,Int64}}, ::Tuple{Array{Float64,2},Array{Float64,2}}, ::Tuple{Int64}) at /home/leo/jcode/OMEinsum.jl/src/einsum.jl:49
 [3] top-level scope at none:0
```

and this one
```julia
julia> einsum(((1,2), (1,3), (2,3,4)), (a, a, b), (1,4))
ERROR: ArgumentError: no valid permutation of dimensions
Stacktrace:
 [1] permutedims(::Array{Float64,1}, ::Tuple{Int64,Int64}) at ./multidimensional.jl:1309
 [2] einsum(::Tuple{Tuple{Int64,Int64},Tuple{Int64,Int64},Tuple{Int64,Int64,Int64}}, ::Tuple{Array{Float64,2},Array{Float64,2},Array{Float64,3}}, ::Tuple{Int64,Int64}) at /home/leo/jcode/OMEinsum.jl/src/einsum.jl:49
 [3] top-level scope at none:0
```

### PR DESCRIPTION
**Pull request by GiggleLiu**
## Autodiff

Back propagate `einsum` in 10 lines!

run `test/autodiff.jl` for autodiff features.

Note: need to fix #4 first, otherwise 2/3 tests would fail.

## Multiple dispatch
`julia test/traits.jl`

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -1,38 +1,10 @@
...
-    aaa = zeros(2,2,2);
-    for (i,j,k,l) in Iterators.product(1:2,1:2,1:2,1:2)
-        aaa[j,k,l] += a[i,j] * a[i,k] * a[i,l]
-    end
-    @test aaa ≈ einsum(((1,2),(1,3),(1,4)), (a,a,a))
-
-    let p = (2,3,1,4)
-        @test einsum(((1,2,3,4),), (t,),p) ≈ permutedims(t,p)
-    end
+@testset "autodiff" begin
```
Comment (under-Peter) [inline test/runtests.jl]: should be "einsum"

Comment (GiggleLiu) [reply test/runtests.jl]: haha, nice catch.

Comment (GiggleLiu) [reply test/runtests.jl]: But it has been fixed, you are reviewing an outdated branch though.

**Comment by GiggleLiu** [comment]
Note: this branch is not intended for merge.

**Pull request closed without merging**